File: tests/dpop.test.ts
import { afterEach, beforeEach, describe, expect, test } from "bun:test";
import { mkdtempSync, rmSync, statSync, readFileSync, existsSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { createHash } from "node:crypto";
import {
  compactVerify,
  decodeJwt,
  decodeProtectedHeader,
  importJWK,
  type JWK,
} from "jose";
import type { Config } from "../src/config.ts";
import { DpopKeyManager, canonicalHtu, toPublicJwk } from "../src/dpop.ts";

let home: string;

function cfg(overrides: Partial<Config> = {}): Config {
  return {
    ADAPTER_BASE_URL: "https://adapter.example.com",
    OKTA_CLIENT_ID: "cid",
    AGENT_ID: "agent-1",
    OKTA_REDIRECT_PORT: 0,
    OKTA_SCOPES: "openid offline_access",
    DPOP_ALG: "ES256",
    DPOP_KEY_MODE: "persistent",
    BRIDGE_HOME: home,
    HTTP_TIMEOUT_MS: 30000,
    LOG_LEVEL: "error",
    ...overrides,
  };
}

beforeEach(() => {
  home = mkdtempSync(join(tmpdir(), "okta-mcp-bridge-"));
});

afterEach(() => {
  rmSync(home, { recursive: true, force: true });
});

describe("createProof", () => {
  test("header is typ dpop+jwt, alg ES256, with an embedded public JWK", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const proof = await mgr.createProof({ htm: "post", htu: "https://adapter.example.com/" });
    const header = decodeProtectedHeader(proof) as { typ?: string; alg?: string; jwk?: JWK };
    expect(header.typ).toBe("dpop+jwt");
    expect(header.alg).toBe("ES256");
    expect(header.jwk).toBeDefined();
    expect(header.jwk?.kty).toBe("EC");
    expect(header.jwk?.crv).toBe("P-256");
  });

  test("embedded jwk has NO private parameters", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const proof = await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/" });
    const header = decodeProtectedHeader(proof) as { jwk?: Record<string, unknown> };
    for (const param of ["d", "p", "q", "dp", "dq", "qi", "k"]) {
      expect(header.jwk?.[param]).toBeUndefined();
    }
  });

  test("payload has jti, upper-cased htm, canonical htu, and iat", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const before = Math.floor(Date.now() / 1000);
    const proof = await mgr.createProof({
      htm: "post",
      htu: "https://Adapter.Example.com:443/?q=1#frag",
    });
    const payload = decodeJwt(proof);
    expect(typeof payload.jti).toBe("string");
    expect((payload.jti as string).length).toBeGreaterThan(0);
    expect(payload.htm).toBe("POST");
    expect(payload.htu).toBe("https://adapter.example.com/");
    expect(typeof payload.iat).toBe("number");
    expect(payload.iat as number).toBeGreaterThanOrEqual(before);
  });

  test("ath is present and correct when an accessToken is provided", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const token = "access-token-xyz";
    const proof = await mgr.createProof({
      htm: "POST",
      htu: "https://adapter.example.com/",
      accessToken: token,
    });
    const payload = decodeJwt(proof) as { ath?: string };
    const expected = createHash("sha256").update(token, "utf8").digest("base64url");
    expect(payload.ath).toBe(expected);
  });

  test("ath is absent when no accessToken is provided", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const proof = await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/" });
    const payload = decodeJwt(proof) as { ath?: string };
    expect(payload.ath).toBeUndefined();
  });

  test("nonce is included only when provided", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const without = decodeJwt(
      await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/" }),
    ) as { nonce?: string };
    expect(without.nonce).toBeUndefined();

    const withNonce = decodeJwt(
      await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/", nonce: "abc123" }),
    ) as { nonce?: string };
    expect(withNonce.nonce).toBe("abc123");
  });

  test("jti is unique across two proofs", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const a = decodeJwt(await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/" }));
    const b = decodeJwt(await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/" }));
    expect(a.jti).not.toBe(b.jti);
  });

  test("signature verifies against the embedded public key", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const proof = await mgr.createProof({ htm: "POST", htu: "https://adapter.example.com/" });
    const header = decodeProtectedHeader(proof) as { jwk: JWK };
    const publicKey = await importJWK(header.jwk, "ES256");
    const result = await compactVerify(proof, publicKey);
    expect(result.protectedHeader.typ).toBe("dpop+jwt");
  });
});

describe("canonicalHtu", () => {
  test("drops query + fragment, strips :443, lower-cases host", () => {
    expect(canonicalHtu("https://Host.EXAMPLE.com:443/?q=1#f")).toBe("https://host.example.com/");
  });

  test("strips :80 for http and keeps the path", () => {
    expect(canonicalHtu("http://Host.example.com:80/token")).toBe("http://host.example.com/token");
  });

  test("keeps a non-default port", () => {
    expect(canonicalHtu("https://host.example.com:8443/cb?x=1")).toBe(
      "https://host.example.com:8443/cb",
    );
  });
});

describe("toPublicJwk", () => {
  test("removes all private parameters", () => {
    const full: JWK = { kty: "EC", crv: "P-256", x: "X", y: "Y", d: "SECRET" };
    const pub = toPublicJwk(full);
    expect(pub.d).toBeUndefined();
    expect(pub.x).toBe("X");
    expect(pub.y).toBe("Y");
  });
});

describe("key persistence", () => {
  test("persistent mode: a second manager loads the SAME jkt from disk", async () => {
    const first = await DpopKeyManager.create(cfg());
    const second = await DpopKeyManager.create(cfg());
    expect(await second.jkt()).toBe(await first.jkt());
  });

  test("ephemeral mode: a second manager has a DIFFERENT jkt", async () => {
    const first = await DpopKeyManager.create(cfg({ DPOP_KEY_MODE: "ephemeral" }));
    const second = await DpopKeyManager.create(cfg({ DPOP_KEY_MODE: "ephemeral" }));
    expect(await second.jkt()).not.toBe(await first.jkt());
  });

  test("dpop-key.json and .seed are chmod 600", async () => {
    await DpopKeyManager.create(cfg());
    const keyMode = statSync(join(home, "dpop-key.json")).mode & 0o777;
    const seedMode = statSync(join(home, ".seed")).mode & 0o777;
    expect(keyMode).toBe(0o600);
    expect(seedMode).toBe(0o600);
  });

  test("the key file is not the plaintext exported JWK", async () => {
    const mgr = await DpopKeyManager.create(cfg());
    const raw = readFileSync(join(home, "dpop-key.json"), "utf8");
    // No private 'd' member and no public 'x' coordinate appear in plaintext.
    expect(raw).not.toContain('"d"');
    expect(raw).not.toContain('"x"');
    const parsed = JSON.parse(raw) as Record<string, unknown>;
    expect(parsed.ciphertext).toBeDefined();
    expect(parsed.iv).toBeDefined();
    expect(parsed.tag).toBeDefined();

    // ...and the round-tripped key is usable: same jkt after reload.
    const reloaded = await DpopKeyManager.create(cfg());
    expect(await reloaded.jkt()).toBe(await mgr.jkt());
    expect(existsSync(join(home, "dpop-key.json"))).toBe(true);
  });
});

File: internal/dpop/dpop_test.go
package dpop

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"math/big"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	"github.com/joewitt99/bridge-mcp-client/internal/config"
)

func testCfg(home, mode string) config.Config {
	return config.Config{
		AdapterBaseURL: "https://adapter.example.com",
		OktaClientID:   "cid",
		AgentID:        "agent-1",
		OktaScopes:     "openid offline_access",
		DpopAlg:        "ES256",
		DpopKeyMode:    mode,
		BridgeHome:     home,
		HTTPTimeout:    30 * time.Second,
		LogLevel:       "error",
	}
}

func newKM(t *testing.T, mode string) *KeyManager {
	t.Helper()
	km, err := NewKeyManager(testCfg(t.TempDir(), mode), nil)
	if err != nil {
		t.Fatal(err)
	}
	return km
}

func mustProof(t *testing.T, km *KeyManager, o ProofOptions) string {
	t.Helper()
	p, err := km.CreateProof(o, nil)
	if err != nil {
		t.Fatal(err)
	}
	return p
}

func decodeSeg(t *testing.T, s string) []byte {
	t.Helper()
	b, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		t.Fatalf("base64url decode: %v", err)
	}
	return b
}

func parseProof(t *testing.T, proof string) (hdr, pl map[string]any, parts []string) {
	t.Helper()
	parts = strings.Split(proof, ".")
	if len(parts) != 3 {
		t.Fatalf("expected 3 JWT segments, got %d", len(parts))
	}
	if err := json.Unmarshal(decodeSeg(t, parts[0]), &hdr); err != nil {
		t.Fatalf("header: %v", err)
	}
	if err := json.Unmarshal(decodeSeg(t, parts[1]), &pl); err != nil {
		t.Fatalf("payload: %v", err)
	}
	return hdr, pl, parts
}

func TestProofHeaderAndClaims(t *testing.T) {
	km := newKM(t, "persistent")
	before := time.Now().Unix()
	hdr, pl, _ := parseProof(t, mustProof(t, km, ProofOptions{
		HTM: "post",
		HTU: "https://Adapter.Example.com:443/?q=1#frag",
	}))

	if hdr["typ"] != "dpop+jwt" || hdr["alg"] != "ES256" {
		t.Fatalf("bad header: %v", hdr)
	}
	jwk, ok := hdr["jwk"].(map[string]any)
	if !ok {
		t.Fatal("header missing jwk")
	}
	if jwk["kty"] != "EC" || jwk["crv"] != "P-256" {
		t.Fatalf("bad jwk: %v", jwk)
	}
	if _, leaked := jwk["d"]; leaked {
		t.Fatal("embedded jwk leaked the private parameter d")
	}
	if s, _ := pl["jti"].(string); s == "" {
		t.Fatal("missing jti")
	}
	if pl["htm"] != "POST" {
		t.Fatalf("htm not upper-cased: %v", pl["htm"])
	}
	if pl["htu"] != "https://adapter.example.com/" {
		t.Fatalf("htu not canonical: %v", pl["htu"])
	}
	iat, ok := pl["iat"].(float64)
	if !ok || int64(iat) < before {
		t.Fatalf("bad iat: %v", pl["iat"])
	}
}

func TestAth(t *testing.T) {
	km := newKM(t, "persistent")
	token := "access-token-xyz"
	_, pl, _ := parseProof(t, mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/", AccessToken: token}))
	sum := sha256.Sum256([]byte(token))
	want := base64.RawURLEncoding.EncodeToString(sum[:])
	if pl["ath"] != want {
		t.Fatalf("ath = %v, want %s", pl["ath"], want)
	}

	_, pl2, _ := parseProof(t, mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/"}))
	if _, ok := pl2["ath"]; ok {
		t.Fatal("ath should be absent without an access token")
	}
}

func TestNonceOnlyWhenProvided(t *testing.T) {
	km := newKM(t, "persistent")
	_, plNo, _ := parseProof(t, mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/"}))
	if _, ok := plNo["nonce"]; ok {
		t.Fatal("nonce should be absent")
	}
	_, plYes, _ := parseProof(t, mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/", Nonce: "abc123"}))
	if plYes["nonce"] != "abc123" {
		t.Fatalf("nonce = %v", plYes["nonce"])
	}
}

func TestJtiUnique(t *testing.T) {
	km := newKM(t, "persistent")
	_, a, _ := parseProof(t, mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/"}))
	_, b, _ := parseProof(t, mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/"}))
	if a["jti"] == b["jti"] {
		t.Fatal("jti should be unique across proofs")
	}
}

func TestSignatureVerifies(t *testing.T) {
	km := newKM(t, "persistent")
	proof := mustProof(t, km, ProofOptions{HTM: "POST", HTU: "https://adapter.example.com/"})
	hdr, _, parts := parseProof(t, proof)
	jwk := hdr["jwk"].(map[string]any)
	pub := &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     new(big.Int).SetBytes(decodeSeg(t, jwk["x"].(string))),
		Y:     new(big.Int).SetBytes(decodeSeg(t, jwk["y"].(string))),
	}
	sig := decodeSeg(t, parts[2])
	if len(sig) != 64 {
		t.Fatalf("ES256 signature must be 64 bytes (R||S), got %d", len(sig))
	}
	digest := sha256.Sum256([]byte(parts[0] + "." + parts[1]))
	if !ecdsa.Verify(pub, digest[:], new(big.Int).SetBytes(sig[:32]), new(big.Int).SetBytes(sig[32:])) {
		t.Fatal("signature did not verify against the embedded public key")
	}
}

func TestCanonicalHTU(t *testing.T) {
	cases := map[string]string{
		"https://Host.EXAMPLE.com:443/?q=1#f":     "https://host.example.com/",
		"http://Host.example.com:80/token":        "http://host.example.com/token",
		"https://host.example.com:8443/cb?x=1":    "https://host.example.com:8443/cb",
		"https://JOE.oktapreview.com/oauth2/v1/x": "https://joe.oktapreview.com/oauth2/v1/x",
	}
	for in, want := range cases {
		got, err := CanonicalHTU(in)
		if err != nil {
			t.Fatalf("%s: %v", in, err)
		}
		if got != want {
			t.Errorf("CanonicalHTU(%q) = %q, want %q", in, got, want)
		}
	}
	if _, err := CanonicalHTU("not a url"); err == nil {
		t.Error("expected error for non-absolute URL")
	}
}

func TestPersistentSameJKT(t *testing.T) {
	home := t.TempDir()
	a, err := NewKeyManager(testCfg(home, "persistent"), nil)
	if err != nil {
		t.Fatal(err)
	}
	b, err := NewKeyManager(testCfg(home, "persistent"), nil)
	if err != nil {
		t.Fatal(err)
	}
	if a.JKT() != b.JKT() {
		t.Fatalf("persistent reload changed jkt: %s != %s", a.JKT(), b.JKT())
	}
}

func TestEphemeralDifferentJKT(t *testing.T) {
	a := newKM(t, "ephemeral")
	b := newKM(t, "ephemeral")
	if a.JKT() == b.JKT() {
		t.Fatal("ephemeral managers should have different jkts")
	}
}

func TestKeyFileSealedAndMode(t *testing.T) {
	home := t.TempDir()
	if _, err := NewKeyManager(testCfg(home, "persistent"), nil); err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(home, "dpop-key.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(raw), `"d"`) || strings.Contains(string(raw), `"x"`) {
		t.Fatal("key file should not contain plaintext private/public JWK params")
	}
	if !strings.Contains(string(raw), `"ct"`) {
		t.Fatal("key file should contain the sealed ciphertext")
	}
	if runtime.GOOS != "windows" {
		fi, err := os.Stat(path)
		if err != nil {
			t.Fatal(err)
		}
		if fi.Mode().Perm() != 0o600 {
			t.Errorf("dpop-key.json mode = %o, want 600", fi.Mode().Perm())
		}
	}
}
